backup: TestCloudBackupRestore_FlakyStorage failed

backup.TestCloudBackupRestore_FlakyStorage [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_CloudUnitTests/21325225?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_CloudUnitTests/21325225?buildTab=artifacts#/) on master @ [ce133afd4b8a7407a72eafb9d6934a900d427d10](https://github.com/cockroachdb/cockroach/commits/ce133afd4b8a7407a72eafb9d6934a900d427d10):


```
=== RUN   TestCloudBackupRestore_FlakyStorage
    test_log_scope.go:172: test logs captured to: /artifacts/tmp/_tmp/91f9cd9ab8ebb37573b189349d153226/logTestCloudBackupRestore_FlakyStorage3620013369
    test_log_scope.go:82: use -show-logs to present logs inline
    test_server_shim.go:637: DRPC is enabled (via testserver factory defaults)
    testutils.go:368: no Invalid Descriptors
    flaky_storage_test.go:130: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/91f9cd9ab8ebb37573b189349d153226/logTestCloudBackupRestore_FlakyStorage3620013369
--- FAIL: TestCloudBackupRestore_FlakyStorage (163.66s)
=== RUN   TestCloudBackupRestore_FlakyStorage/cloud=gcp
    flaky_storage_test.go:85: using backup store URI: gs://nightly-cloud-unit-tests/TestCloudBackupRestore_FlakyStorage%252Fcloud=gcp-1776989807143228380?AUTH=implicit
    flaky_storage_test.go:119: condition failed to evaluate within 45s: pq: restore job exhausted max retries without making progress: importing 1 ranges: running distributed restore: running distSQL flow: reading restore span entries: making backup metadata: failed to read backup descriptor: injected error for op 'faultyReader.Read' on object 'BACKUP_METADATA'
        
        goroutine dump: /artifacts/tmp/_tmp/91f9cd9ab8ebb37573b189349d153226/goroutine_dump_114436587.txt
--- FAIL: TestCloudBackupRestore_FlakyStorage/cloud=gcp (117.65s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
/cc @cockroachdb/disaster-recovery
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestCloudBackupRestore_FlakyStorage.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>


Jira issue: CRDB-63219

backup.TestCloudBackupRestore_FlakyStorage [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_CloudUnitTests/21329029?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_CloudUnitTests/21329029?buildTab=artifacts#/) on master @ [3399ae7c9d6e7b7c840a2355f9fbc50a4cf28ffe](https://github.com/cockroachdb/cockroach/commits/3399ae7c9d6e7b7c840a2355f9fbc50a4cf28ffe):


```
=== RUN   TestCloudBackupRestore_FlakyStorage
    test_log_scope.go:172: test logs captured to: /artifacts/tmp/_tmp/91f9cd9ab8ebb37573b189349d153226/logTestCloudBackupRestore_FlakyStorage3754889505
    test_log_scope.go:82: use -show-logs to present logs inline
    test_server_shim.go:637: DRPC is enabled (via testserver factory defaults)
    testutils.go:368: no Invalid Descriptors
    flaky_storage_test.go:130: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/91f9cd9ab8ebb37573b189349d153226/logTestCloudBackupRestore_FlakyStorage3754889505
--- FAIL: TestCloudBackupRestore_FlakyStorage (155.86s)
=== RUN   TestCloudBackupRestore_FlakyStorage/cloud=gcp
    flaky_storage_test.go:85: using backup store URI: gs://nightly-cloud-unit-tests/TestCloudBackupRestore_FlakyStorage%252Fcloud=gcp-1777335275495945229?AUTH=implicit
    flaky_storage_test.go:119: condition failed to evaluate within 45s: pq: failed to read backup from "gs://nightly-cloud-unit-tests/TestCloudBackupRestore_FlakyStorage%252Fcloud=gcp-1777335275495945229/incrementals/2026/04/28-001439.66/20260428/001451.24-20260428-001447.62?AUTH=implicit": externalstorage.read failed: injected error for 'BACKUP_METADATA-CHECKSUM'
        
        goroutine dump: /artifacts/tmp/_tmp/91f9cd9ab8ebb37573b189349d153226/goroutine_dump_3145674991.txt
--- FAIL: TestCloudBackupRestore_FlakyStorage/cloud=gcp (107.40s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestCloudBackupRestore_FlakyStorage.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>

backup.TestCloudBackupRestore_FlakyStorage [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_CloudUnitTests/21330942?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_CloudUnitTests/21330942?buildTab=artifacts#/) on master @ [fac65e5230ed33ed99819f22fd14c08c31785079](https://github.com/cockroachdb/cockroach/commits/fac65e5230ed33ed99819f22fd14c08c31785079):


```
=== RUN   TestCloudBackupRestore_FlakyStorage
    test_log_scope.go:172: test logs captured to: /artifacts/tmp/_tmp/91f9cd9ab8ebb37573b189349d153226/logTestCloudBackupRestore_FlakyStorage1990306516
    test_log_scope.go:82: use -show-logs to present logs inline
    test_server_shim.go:168: automatically injected an external process virtual cluster under test; see comment at top of test_server_shim.go for details.
    testutils.go:368: no Invalid Descriptors
    flaky_storage_test.go:130: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/91f9cd9ab8ebb37573b189349d153226/logTestCloudBackupRestore_FlakyStorage1990306516
--- FAIL: TestCloudBackupRestore_FlakyStorage (179.43s)
=== RUN   TestCloudBackupRestore_FlakyStorage/cloud=gcp
    flaky_storage_test.go:85: using backup store URI: gs://nightly-cloud-unit-tests/TestCloudBackupRestore_FlakyStorage%252Fcloud=gcp-1777421743588888868?AUTH=implicit
    flaky_storage_test.go:119: condition failed to evaluate within 45s: pq: restore job exhausted max retries without making progress: importing 1 ranges: running distributed restore: running distSQL flow: reading restore span entries: making backup metadata: failed to read backup descriptor: externalstorage.read failed: injected error for 'BACKUP_METADATA'
        
        goroutine dump: /artifacts/tmp/_tmp/91f9cd9ab8ebb37573b189349d153226/goroutine_dump_2614359388.txt
--- FAIL: TestCloudBackupRestore_FlakyStorage/cloud=gcp (104.54s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestCloudBackupRestore_FlakyStorage.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>

--- a/pkg/backup/flaky_storage_test.go
+++ b/pkg/backup/flaky_storage_test.go
@@ -11,6 +11,7 @@ import (
 	"net/url"
 	"os"
 	"testing"
+	"time"
 
 	"github.com/aws/aws-sdk-go-v2/config"
 	"github.com/cockroachdb/cockroach/pkg/backup/backuptestutils"
@@ -84,6 +85,13 @@ func TestCloudBackupRestore_FlakyStorage(t *testing.T) {
 			backupURI := uri.String()
 			t.Logf("using backup store URI: %s", backupURI)
 
+			// Cloud storage operations are slower due to network latency,
+			// so use a longer timeout than the default SucceedsSoon duration.
+			successTimeout := testutils.DefaultSucceedsSoonDuration
+			if store.Cloud() != "nodelocal" {
+				successTimeout = 90 * time.Second
+			}
+
 			nextRow := 1
 			writeRows := func(t *testing.T, sql *sqlutils.SQLRunner) {
 				for i := 0; i < 10; i++ {
@@ -97,31 +105,31 @@ func TestCloudBackupRestore_FlakyStorage(t *testing.T) {
 			sql.Exec(t, `CREATE TABLE testdb.test_table (id INT PRIMARY KEY, name STRING, value INT)`)
 
 			writeRows(t, sql)
-			testutils.SucceedsSoon(t, func() error {
+			testutils.SucceedsWithin(t, func() error {
 				_, err := sqlDB.DB.ExecContext(ctx, `BACKUP TABLE testdb.test_table INTO $1`, backupURI)
 				return err
-			})
+			}, successTimeout)
 
 			for i := 0; i < 10; i++ {
 				writeRows(t, sql)
-				testutils.SucceedsSoon(t, func() error {
+				testutils.SucceedsWithin(t, func() error {
 					_, err := sqlDB.DB.ExecContext(
 						ctx, `BACKUP TABLE testdb.test_table INTO LATEST IN $1`, backupURI,
 					)
 					return err
-				})
+				}, successTimeout)
 			}
 
 			originalFingerprint := sql.QueryStr(t, `SHOW EXPERIMENTAL_FINGERPRINTS FROM TABLE testdb.test_table`)
 
 			sqlDB.Exec(t, `DROP TABLE testdb.test_table`)
 
-			testutils.SucceedsSoon(t, func() error {
+			testutils.SucceedsWithin(t, func() error {
 				_, err := sqlDB.DB.ExecContext(
 					ctx, `RESTORE TABLE testdb.test_table FROM LATEST IN $1`, backupURI,
 				)
 				return err
-			})
+			}, successTimeout)
 
 			restoredFingerprint := sqlDB.QueryStr(t, `SHOW EXPERIMENTAL_FINGERPRINTS FROM TABLE testdb.test_table`)
 			require.Equal(t, originalFingerprint, restoredFingerprint, "fingerprints should match")

